### ISSUE TITLE
Underscore not allowed in attribute names

### ISSUE THREAD

**Issue by chuckdumont** [OWASP/java-html-sanitizer#120]
Why is the underscore character `_` not allowed in attribute names?  I can't find anything in the html spec that disallows it.  Even if I whitelist the name, it's still doesn't make it though to the output.  It seems that `org.owasp.html.HtmlStreamRenderer.isValidHtmlName(String name)` is blocking any name (element or attribute) that has an underscore in it and there doesn't seem to be any way to configure this behavior.

**Comment by mikesamuel** [comment]
TLDR, send me a PR to `isValidHtmlName` and I'll approve it.

I tried to answer your questions to refresh my memories, so below is mostly me talking to myself, but feel free to read if you want a history lesson.

----

> there doesn't seem to be any way to configure this behavior.

There is no way to configure it.

>  I can't find anything in the html spec that disallows it.

Per the spec, I think the relevant part is [12.2.5.33 Attribute name state](https://html.spec.whatwg.org/#attribute-name-state)
which define an attribute name as roughly `/[^\t\n\f\r >="'<]+/` with the caveats that `[A-Z]` are lowercased and `[\0]` is converted to U+FFFD.

So you're right that you won't get a parse error.  You will of course get a schema error like "attribute `foo_bar` not allowed on element ..." during tree building.

> Why is the underscore character _ not allowed in attribute names?

I was probably being overly cautious at the time I wrote that.

The [relevant bit](https://github.com/OWASP/java-html-sanitizer/blob/master/src/main/java/org/owasp/html/HtmlLexer.java#L130-L140) of the parser should handle the underscores just fine.

You were right about the root cause bing [`isValidHtmlName` call](https://github.com/OWASP/java-html-sanitizer/blob/master/src/main/java/org/owasp/html/HtmlStreamRenderer.java#L191).

Most of the security properties that aren't specific to a particular policy are enforced in the HTML renderer.
IIRC I designed the renderer to produce a consistently parsed subset of HTML to avoid triggering corner cases in browsers parsers.  My best guess at "consistently parsed subset of HTML" was HTML that looks a lot like XHTML.  `isValidHtmlName` was defined to be the intersection of HTML element name, HTML attribute name, and XML identifier, and possibly CSS identifier.  I think the latter two (CSS definitely) initially disallowed underscore though I think that changed.

**Comment by chuckdumont** [comment]
I submitted PR #127 but the CI builds failed due to issues that I don't believe are related to my change.
